### PR TITLE
feat(scraper): add possibility to remove labels

### LINKED ISSUE #86: Feature: Remove label during scraping
I have a specific usecase where a prometheus endpoint is generating useless labels that are creating many GTS, I would love to drop labels during scraping ❤️ 

Usecase:

Flink is generating metrics with a `jobname` and `jobid` label. JobID is changing every time the job is restarted. I'm only querying through the jobname label.

### PR DESCRIPTION
**Pull request by PierreZ**
Resolve #86   ❤️

### CODE REVIEW

File: src/scraper.rs [hunk 1]
```diff
@@ -140,6 +140,14 @@ fn fetch(
                 }
             };
 
+            let line = match lib::remove_labels(&line, &scraper.filtered_labels) {
```
Comment (FlorentinDUBOIS) [inline src/scraper.rs]: check before the match that `filtered_labels` is empty

File: src/lib/mod.rs [hunk 2]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
 
     Ok(format!("{}{{{}{}{}", class, labels, sep, plabels))
 }
+
+pub fn remove_labels(line: &str, labels_to_drop: &Vec<String>) -> Result<String, Box<Error>> {
+    if labels_to_drop.is_empty() {
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]: remove this check from here

File: src/lib/mod.rs [hunk 3]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+    let plabels = end_parts.next().ok_or("no_end")?;
+    let plabels = String::from(plabels);
+
+    let value = end_parts.next().ok_or("no_value")?;
+    let value = String::from(value);
+
+    let labels: Vec<String> = plabels
+        .split(",")
+        .filter_map(|l| {
+            let label_splits: Vec<&str> = l.split("=").collect();
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]:
```rust
let label_splits: Vec<String> = l.split("=").map(|s| String::from(s)).collect();
```

File: src/lib/mod.rs [hunk 4]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+            let key = label_splits.get(0)?;
+            let key = String::from(*key);
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]: remove this line

File: src/lib/mod.rs [hunk 5]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+
+            let value = label_splits.get(1)?;
+            let value = String::from(*value);
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]: remove this line

File: src/lib/mod.rs [hunk 6]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+            Some((key, value))
+        })
+        .filter(|(key, _value)| !labels_to_drop.contains(&key.to_owned()))
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]:
```rust
.filter(|(ref key, _)| !labels_to_drop.contains(key))
```
I wonder if this works ?

File: src/lib/mod.rs [hunk 7]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn remove_no_labels() {
+        let line = "1484828198557102// f{job_id=123,job_name=job1,another_id=456} 10";
+        let expected: Result<String, Box<Error>> = Ok(String::from(line));
+        let labels = Vec::new();
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]: vec![] is shorter ;)

File: src/lib/mod.rs [hunk 8]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+        .filter(|(key, _value)| !labels_to_drop.contains(&key.to_owned()))
+        .map(|(key, value)| format!("{}={}", key, value))
+        .collect();
+
+    Ok(format!("{}{{{}}} {}", class, labels.join(","), value))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]: Is it useful ?

File: src/lib/mod.rs [hunk 9]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+        let result = super::remove_labels(line, &labels);
+        assert_eq!(expected.is_ok(), result.is_ok());
+        assert_eq!(expected.unwrap(), result.unwrap());
+    }
+    #[test]
+    fn remove_multiple_labels() {
+        let line = "1484828198557102// f{job_id=123,job_name=job1,another_id=456} 10";
+        let expected: Result<String, Box<Error>> =
+            Ok(String::from("1484828198557102// f{job_id=123} 10"));
+        let mut labels = Vec::new();
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]:
```rust
let labels = vec![
"job_name",
"another_id",
];
```

File: src/lib/mod.rs [hunk 10]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+        labels.push(String::from("job_name"));
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]: remove this line

File: src/lib/mod.rs [hunk 11]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+        labels.push(String::from("another_id"));
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]: remove this line

File: src/lib/mod.rs [hunk 12]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+        assert_eq!(expected.unwrap(), result.unwrap());
+    }
+    #[test]
+    fn remove_one_labels() {
+        let line = "1484828198557102// f{job_id=123,job_name=job1,another_id=456} 10";
+        let expected: Result<String, Box<Error>> = Ok(String::from(
+            "1484828198557102// f{job_id=123,another_id=456} 10",
+        ));
+        let mut labels = Vec::new();
+        labels.push(String::from("job_name"));
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]: remove this line

File: src/lib/mod.rs [hunk 13]
```diff
@@ -23,3 +23,81 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
+        assert_eq!(expected.is_ok(), result.is_ok());
+        assert_eq!(expected.unwrap(), result.unwrap());
+    }
+    #[test]
+    fn remove_one_labels() {
+        let line = "1484828198557102// f{job_id=123,job_name=job1,another_id=456} 10";
+        let expected: Result<String, Box<Error>> = Ok(String::from(
+            "1484828198557102// f{job_id=123,another_id=456} 10",
+        ));
+        let mut labels = Vec::new();
```
Comment (FlorentinDUBOIS) [inline src/lib/mod.rs]:
```rust
let labels = vec!["job_name"];
```

File: src/scraper.rs (hunk 1 above)
Comment (PierreZ) [reply src/scraper.rs]: Followed design from above to be consistent between add and remove_labels. add_labels does not check if labels is empty

File: src/lib/mod.rs (hunk 3 above)
Comment (PierreZ) [reply src/lib/mod.rs]: love it!

File: src/lib/mod.rs [hunk 14]
```diff
@@ -23,3 +23,75 @@ pub fn add_labels(line: &str, labels: &str) -> Result<String, Box<Error>> {
...
 }
+
+pub fn remove_labels(line: &str, labels_to_drop: &Vec<String>) -> Result<String, Box<Error>> {
+    if labels_to_drop.is_empty() {
+        return Ok(String::from(line));
+    }
+
+    let mut parts = line.splitn(2, '{');
+
+    let class = parts.next().ok_or("no_class")?;
```
Comment (d33d33) [inline src/lib/mod.rs]: ok_or ?

**Pull request merged**